### ISSUE TITLE
Order of zones on splitter/mixer causing diffs

### ISSUE THREAD

**Issue by macumber** [NREL/EnergyPlus#6508]
Issue overview
--------------
When testing OpenStudio for regression changes, we noticed that the following files produce different simulation results.  As far as I know, the only real difference here is the order of zones on the splitter/mixer.  In my understanding this should not cause simulation diffs.  The files were reported by @jmarrec, @kbenne has not had a chance to review these yet.

### Details
Some additional details for this issue (if relevant):
 - Ubuntu
 - EnergyPlus 8.8 official

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by macumber** [comment]
[centralheatpumpsystem.rb_2.4.2_out_Ubuntu_run2_sorted.idf.txt](https://github.com/NREL/EnergyPlus/files/1776075/centralheatpumpsystem.rb_2.4.2_out_Ubuntu_run2_sorted.idf.txt)
[centralheatpumpsystem.rb_2.4.2_out_Ubuntu_run8_sorted.idf.txt](https://github.com/NREL/EnergyPlus/files/1776074/centralheatpumpsystem.rb_2.4.2_out_Ubuntu_run8_sorted.idf.txt)

**Comment by mjwitte** [comment]
Hmmm, the changes are on the order of 1% at the ABUPS level, but one would expect small diffs in the decimals for this type of change. Original files have 3 changes - different order of East/West nodes in the ZoneMixer and the ZoneSplitter plus different order of the East/West reheat coil branches in the plant loop.  This zip has 5 idfs updated to v8.9.0 - separating out the three changes hoping to isolate the culprit. Instead, got five different answers. @rraustad @Myoldmopar 
[6508-Order dependence-v890-MoreVariations.zip](https://github.com/NREL/EnergyPlus/files/1804111/6508-Order.dependence-v890-MoreVariations.zip)

Ran these files with BranchPumps branch, v8.9.0 IDD Freeze, and with a branch checked out with the last merge before the JSON input refactor - all show similar diffs. And then (duh) ran these with v8.8 and similar diffs are there, too. So this is not something new introduced in v8.9. These files have lots of plant warnings, so that may be adding some instability.

**Comment by jmarrec** [comment]
@mjwitte we have 20 tests that are based on the same basis that creates an ASHRAE sys 7 (VAV w/reheat), then we replace/add stuff on top of that. There are 4 additional that are system type 8. 

The base case is [baseline_sys07.rb](https://github.com/jmarrec/OpenStudio-resources/blob/develop/model/simulationtests/baseline_sys07.rb). Any tests that share this same basis is subject to variations in the order of thermal zones.

Here the only thing I'm doing differently is to replace the boiler and the chiller with a CentralHeatPumpSystem. 

None of the other tests exhibit such differences in ABUPS, which makes me think there is possibility it could be related to the central heat pump system.

Here are five runs of baseline_sys07.rb, percentages computed compared to Run1:

![image](https://user-images.githubusercontent.com/5479063/37711209-3e13c17a-2d10-11e8-804f-e8c6729b30bc.png)

**Comment by mjwitte** [comment]
@jmarrec I agree, it looks like the problem is the central heat pump system, but that may just be a contributing factor. An interesting test for the files with the heat pump system would be to add a `ConvergenceLimits` object and increase the "Minimum Plant Iterations" to, say 4 or 5. Perhaps the problem is that the central heat pump needs more iterations to update everything and converge on a solution.

**Comment by jmarrec** [comment]
Here are 5 runs with the convergencelimits as you requested:

![image](https://user-images.githubusercontent.com/5479063/37714789-b4026e40-2d1a-11e8-9426-8e2f37f6b139.png)


```
ConvergenceLimits,
  1,                                      !- Minimum System Timestep {minutes}
  ,                                       !- Maximum HVAC Iterations
  5;                                      !- Minimum Plant Iterations
```

**Comment by mjwitte** [comment]
So, that doesn't solve the problem. Did results change with the extra iterations?

**Comment by jmarrec** [comment]
I'm using the ruby test directly and simply running it several times and storing the site kbtu (I store the out.osw from OpenStudio_results measure actually), as this is simply the easiest for me, but it means I can't predict the order of the zones and I currently don't check for it.

All I can say is that it didn't improve the experienced variation in total site energy.

**Comment by EnergyArchmage** [comment]
We have seen this issue in a model with no plant at all.  Only difference is order in AirLoopHVAC:ZoneMixer object.  Problem is different Fans:Electricity output depending on order.

**Comment by rraustad** [comment]
This is interesting. Adding @Myoldmopar @nealkruis @jasondegraw @JasonGlazer @Nigusse @lgu1234 

Using a different test file (air loop only with no plant) the only difference is that the order of the mixer legs has been switched. I still need to try this solution with the 6508 defect files. I also wonder if the plenum has this same problem.

From:

    AirLoopHVAC:ZoneMixer,
        SYS0Demand Mixer, !- Name
        SYS0Demand Outlet Node,  !- Outlet Node Name
        Zone_00_01Return Node,  !- Inlet 1 Node Name
        Zone_00_02Return Node;  !- Inlet 2 Node Name

To:

    AirLoopHVAC:ZoneMixer,
        SYS0Demand Mixer, !- Name
        SYS0Demand Outlet Node,  !- Outlet Node Name
        Zone_00_02Return Node,  !- Inlet 1 Node Name
        Zone_00_01Return Node;  !- Inlet 2 Node Name

The mixer object looks to mass flow average the mixer's various inlet node conditions at the single outlet node and feed those conditions back to the inlet of the air loop. When flow through the mixer exists the mass balanced outlet condition is calculated like this for each parameter, Humrat, Temp, Pressure, and Enthalpy. However, when mass flow through the mixer == 0, there is no way to mass balance the outlet conditions and the code simply takes the first leg of the mixer as if it doesn't matter what that condition is with regards to the simulation results. First, there are valid reasons for 0 return flow (e.g., air loop off or zone exhaust fans) and second, if the input file objects are written arbitrarily and the mixer legs are switched, the results change because now leg (1) points to a different zone (with different T, w and H). If the air loop is actually on, this can make a big difference in results. It may be better to average the inlet nodes when flow = 0 or find a way to verify if the air loop is actually off before a simple average is used. 

        if (MixerCond(MixerNum).OutletMassFlowRate > 0.0) {
            // Mass balance on moisture to get outlet air humidity ratio
            for (InletNodeNum = 1; InletNodeNum <= MixerCond(MixerNum).NumInletNodes; ++InletNodeNum) {
                MixerCond(MixerNum).OutletHumRat += MixerCond(MixerNum).InletMassFlowRate(InletNodeNum) *
                                                    MixerCond(MixerNum).InletHumRat(InletNodeNum) / MixerCond(MixerNum).OutletMassFlowRate;
            } -- snip out rest of code when mass > 0 --
        } else {
            // Mass Flow in air loop is zero and loop is not operating.
            // Arbitrarily set the output to the first inlet leg
            MixerCond(MixerNum).OutletHumRat = MixerCond(MixerNum).InletHumRat(1);
            MixerCond(MixerNum).OutletPressure = MixerCond(MixerNum).InletPressure(1);
            MixerCond(MixerNum).OutletEnthalpy = MixerCond(MixerNum).InletEnthalpy(1);
            MixerCond(MixerNum).OutletTemp = MixerCond(MixerNum).InletTemp(1);
        }

Verified that if the inlet nodes are average when flow = 0, the results are identical. Does this sound like a reasonable solution?

        } else {
            // Mass Flow in air loop is zero and loop is not operating.
            // Arbitrarily average the inlet legs
            for ( InletNodeNum = 1; InletNodeNum <= MixerCond(MixerNum).NumInletNodes; ++InletNodeNum ) {
                MixerCond(MixerNum).OutletHumRat += MixerCond(MixerNum).InletHumRat(InletNodeNum) / MixerCond(MixerNum).NumInletNodes;
                MixerCond(MixerNum).OutletPressure += MixerCond(MixerNum).InletPressure(InletNodeNum) / MixerCond(MixerNum).NumInletNodes;
                MixerCond(MixerNum).OutletEnthalpy += MixerCond(MixerNum).InletEnthalpy(InletNodeNum) / MixerCond(MixerNum).NumInletNodes;
                MixerCond(MixerNum).OutletTemp += MixerCond(MixerNum).InletTemp(InletNodeNum) / MixerCond(MixerNum).NumInletNodes;
            }
        }

**Comment by rraustad** [comment]
For the defect files provided in the description the AirloopHVAC:ZoneMixer averaging when flow = 0 did not resolve the diff's. I see though that in these files the plant branches are also switched. 

![image](https://user-images.githubusercontent.com/5778156/58107981-f8ce9d80-7bb8-11e9-8c4f-6053b01ea5bb.png)
